python-client/giskard/models/cache/cache.py

```diff
@@ -29,29 +30,45 @@ class ModelCache:
 
     vectorized_get_cache_or_na = None
 
-    def __init__(self, model_type: SupportedModelTypes, id: Optional[str] = None, cache_dir: Path = None):
+    def __init__(self, model_type: SupportedModelTypes, id: Optional[str] = None, cache_dir: Optional[Path] = None):
         self.id = id or str(uuid.uuid4())
         self.prediction_cache = dict()
-        self.cache_dir = cache_dir or Path(settings.home_dir / settings.cache_dir / "global/prediction_cache" / self.id)
-
-        if id is not None:
-            if (self.cache_dir / CACHE_CSV_FILENAME).exists():
-                with open(self.cache_dir / CACHE_CSV_FILENAME, "r") as pred_f:
-                    reader = csv.reader(pred_f)
-                    for row in reader:
-                        if model_type == SupportedModelTypes.TEXT_GENERATION:
-                            self.prediction_cache[row[0]] = row[1:]
-                        elif model_type == SupportedModelTypes.REGRESSION:
-                            self.prediction_cache[row[0]] = float(row[1])
-                        else:
-                            self.prediction_cache[row[0]] = [float(i) for i in row[1:]]
-
+        self.cache_dir = cache_dir or Path(
+            settings.home_dir / settings.cache_dir / "global" / "prediction_cache" / self.id
+        )
+        self.cache_file = self.cache_dir / CACHE_CSV_FILENAME
         self.vectorized_get_cache_or_na = np.vectorize(self.get_cache_or_na, otypes=[object])
+        self.model_type = model_type
+        self._warmed_up = False
+
+    def warm_up_from_disk(self):
+        if id is None or not self.cache_file.exists():
+            return
+
+        try:
+            with self.cache_file.open("r", newline="") as pred_f:
+                reader = csv.reader(pred_f)
+                for row in reader:
+                    if self.model_type == SupportedModelTypes.TEXT_GENERATION:
+                        # Text generation models output should be a single string
+                        self.prediction_cache[row[0]] = row[1]
+                    elif self.model_type == SupportedModelTypes.REGRESSION:
+                        # Regression models output is always casted to float
+                        self.prediction_cache[row[0]] = float(row[1])
+                    else:
+                        # Classification models return list of probabilities
+                        self.prediction_cache[row[0]] = [float(i) for i in row[1:]]
+        except Exception as e:
+            warning(f"Failed to load cache from disk for model {self.id}: {e}")
 
     def get_cache_or_na(self, key: str):
         return self.prediction_cache.get(key, NaN)
 
     def read_from_cache(self, keys: pd.Series):
+        if self.id and not self._warmed_up:
```

Comment:
when can we have a missing `self.id`? There's `self.id = id or str(uuid.uuid4())` in the constructor

Reply:
You are right, I refactored the previous code but I think the generation of new UUID in the constructor should be removed.